src/DictVectors/dvec.jl

```diff
@@ -1,113 +1,156 @@
+  :a => 1
+
+julia> dv = DVec(:a => 2, :b => 3; style=IsDynamicSemistochastic())
+DVec{Symbol,Float32,IsDynamicSemistochastic{true}} with 2 entries
+  :a => 2.0f0
+  :b => 3.0f0
+```
+"""
+struct DVec{K,V,S<:StochasticStyle{V},D<:AbstractDict{K,V}} <: AbstractDVec{K,V}
+    storage::D
+    style::S
 end
 
-# from AbstractDict; note that a new dict is constructed and data is copied
-function DVec(d::AbstractDict{K,V},
-              capacity = length(d)) where K where V
-    dv = DVec{K,V}(capacity)
-    for (k,v) in d
-        dv[k] = v
+###
+### Constructors
+###
+function DVec(args::Vararg{Pair{K,V}}; style=default_style(V), capacity=0) where {K,V}
+    dict = Dict{K,V}()
+    sizehint!(dict, max(length(args), capacity))
+    for (k, v) in args
+        dict[k] = convert(eltype(style), v)
     end
-    return dv
+    return DVec(dict; style)
+end
+# In this constructor, the style matches the dict's valtype.
+function DVec(
+    dict::AbstractDict{K,V}; style::StochasticStyle{V}=default_style(V), capacity=0
+) where {K,V}
+    capacity > 0 && sizehint!(dict, capacity)
+    return DVec{K,V,typeof(style),typeof(dict)}(dict, style)
+end
+# In this constructor, the dict has to be converted to the appropriate valtype.
+function DVec(
+    dict::Dict{K}; style::StochasticStyle{V}=default_style(valtype(dict)), capacity=0
+) where {K,V}
+    dict = convert(Dict{K,V}, dict)
+    return DVec{K,V,typeof(style),typeof(dict)}(dict, style)
+end
+# Constructor from arbitrary iterator
+function DVec(itr; style=nothing, capacity=0)
+    dict = Dict(itr)
+    if isnothing(style)
+        return DVec(dict; capacity)
+    else
+        return DVec(dict; style, capacity)
+    end
+end
+# Empty constructor.
+function DVec{K,V}(; style::StochasticStyle=default_style(V), capacity=0) where {K,V}
+    return DVec(Dict{K,V}(); style, capacity)
+end
+# From another DVec
+function DVec(dv::AbstractDVec{K,V}, style=StochasticStyle(dv), capacity=0) where {K,V}
+    dvec = DVec{K,V}(; style, capacity=max(capacity, length(dv)))
+    return copyto!(dvec, dv)
 end
 
-# from AbstractDVec
-function DVec(adv::AbstractDVec{K,V}, capacity = capacity(adv)) where K where V
-    dv = DVec{K,V}(capacity) # allocate new DVec
-    return copyto!(dv,adv) # generic for AbstractDVec
+function Base.empty(dvec::DVec{K,V}) where {K,V}
+    return DVec{K,V}(; style=StochasticStyle(dvec))
+end
+function Base.empty(dvec::DVec{K}, ::Type{V}) where {K,V}
+    return DVec{K,V}()
+end
+function Base.empty(dvec::DVec, ::Type{K}, ::Type{V}) where {K,V}
+    return DVec{K,V}()
 end
 
-# the following also create and allocated new DVec objects
-Base.empty(dv::DVec{K,V}, c::Integer = capacity(dv) ) where {K,V} = DVec{K,V}(c)
-Base.empty(dv::DVec, ::Type{V}) where {V} = empty(dv,keytype(dv),V)
-Base.empty(dv::DVec, ::Type{K}, ::Type{V}) where {K,V} = DVec{K,V}(capacity(dv))
-Base.similar(dv::DVec, args...) = empty(dv, args...)
+Base.similar(dvec::DVec, args...; kwargs...) = empty(dvec, args...; kwargs...)
```

Comment:
Should `similar` preserve capacity, i.e. should we apply `sizehint!` corresponding to `length(dvec)`? Maybe not.